design-proposals/klip-52-bytes-data-type-support.md

```diff
@@ -0,0 +1,142 @@
+# KLIP 46 - BYTES Data Type Support
+
+**Author**: Zara Lim (@jzaralim) | 
+**Release Target**: 0.21 | 
+**Status**: _In Discussion_ | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/7764
+
+**tl;dr:** _Add support for the BYTES data type. This will allow users to work with BLOBs of data that don't fit into any other data type._
+           
+## Motivation and background
+
+Currently, ksqlDB can only handle a set of primitive types and combinations of them.
+A BYTES data type would allow users to work with data that does not fit into any of
+the primitive types such as images, as well as BLOB/binary data from other databases.
+
+## What is in scope
+* Add BYTES type to KSQL
+* Support BYTE comparisons
+* Support BYTES usage in STRUCT, MAP and ARRAY
+* Serialization and de-serialization of BYTES to Avro, JSON, Protobuf and Delimited formats
+* Adding/updating UDFs to support the BYTES type
+
+## What is not in scope
+* Fixed sized BYTES (`BYTES(3)` representing 3 bytes, for example) - This is supported by Kafka Connect by adding the `connect.fixed.size`
+key in a bytes schema, but this will not be included in this KLIP.
+
+## Public APIS
+
+### BYTES
+
+The BYTES data type will store an array of raw bytes of an unspecified length. The maximum size of
+the array is limited by the maximum size of a Kafka message, as well as possibly by the value format being used.
+The syntax is as follows:
+
+```roomsql
+CREATE STREAM stream_name (b BYTES, COL2 STRING) AS ...
+CREATE TABLE table_name (col1 STRUCT<field BYTES>) AS ...
+```
+
+By default, BYTES will be displayed in console as HEX strings, where each byte is represented by two characters.
+For example, the byte array `[91, 67]` will be displayed as:
+
+```roomsql
+> SELECT b from STREAM;
+'0x5B43'
+```
+
+Implicit conversions to BYTES will not be supported.
+
+### UDF
+
+The following UDFs will be added:
+
+* `to_bytes(string, encoding)` - this will convert a STRING value to BYTES in the specified encoding.
+The accepted encoders are `hex`, `utf8`, `ascii`, and `base64`.
+* `from_bytes(bytes, encoding)` - this will convert a BYTES value to STRING in the specified encoding.
+The accepted encoders are `hex`, `utf8`, `ascii`, and `base64`.
+
+We will also update some of the existing STRING functions to accept BYTES as a parameter. In general, if a function works on ASCII characters for a STRING parameter,
```

Comment:
Will these functions assume a particular encoding for the bytes, or will the functions all need to take a parameter to specify the encoding? If the latter, I wonder if it's (conceptually) simpler to just let users nest the `from_bytes()` function instead.

Reply:
All DBs that support bytes/binary types also support functions to use strings or bytes as parameters. Using `from_bytes()` everytime they want to use a string function will have a different behavior, for instance:

Say we have raw = [91, 67], then:
`len(raw) = 2`, but `len(from_bytes(raw, 'hex')) = 6` because `from_bytes(raw, 'hex') = '0x5B43'`

Reply:
Yup, so basically they just operate on raw bytes, so no encoding is needed.